hs/src/Reach/Simulator.hs

```diff
@@ -0,0 +1,477 @@
+    -- TODO: fake constants per connector
+    DLA_Constant _dlconst -> return $ V_UInt $ toInteger (maxBound :: Int)
+    DLA_Literal dllit -> interp dllit
+    -- TODO: handle interact case: needs participant frontend state
+    DLA_Interact _slpart _string _dltype -> undefined
+
+instance Interp DLLiteral where
+  interp = \case
+    DLL_Null -> return $ V_Null
+    DLL_Bool bool -> return $ V_Bool bool
+    DLL_Int _at int -> return $ V_UInt int
+
+instance Interp DLLargeArg where
+  interp = \case
+    DLLA_Array _dltype dlargs -> V_Array <$> mapM interp dlargs
+    DLLA_Tuple dlargs -> V_Tuple <$> mapM interp dlargs
+    DLLA_Obj strs_to_dlargs -> V_Object <$> mapM interp strs_to_dlargs
+    DLLA_Data _map_slvars_to_dltypes string dlarg -> do
+      evd_arg <- interp $ dlarg
+      return $ V_Data string evd_arg
+    DLLA_Struct assoc_slvars_dlargs -> do
+      evd_args <- mapM (\arg -> interp arg) $ M.fromList assoc_slvars_dlargs
+      return $ V_Struct $ M.toList evd_args
+    DLLA_Bytes _ -> undefined
+
+instance Interp DLExpr where
+  interp = \case
+    DLE_Arg _at dlarg -> interp dlarg
+    DLE_LArg _at dllargearg -> interp dllargearg
+    DLE_Impossible at _int err -> expect_thrown at err
+    DLE_PrimOp _at primop dlargs -> do
+      evd_args <- mapM interp dlargs
+      interpPrim (primop,evd_args)
+    DLE_ArrayRef _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> return $ arr !! (fromIntegral n)
+        _ -> impossible "expression interpreter"
+    DLE_ArraySet _at dlarg1 dlarg2 dlarg3 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      ev3 <- interp dlarg3
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> do
+          let n' = fromIntegral n
+          return $ V_Array $ take n' arr ++ [ev3] ++ drop (n' + 1) arr
+        _ -> impossible "expression interpreter"
+    DLE_ArrayConcat _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr1, V_Array arr2) -> return $ V_Array $ arr1 ++ arr2
+        _ -> impossible "expression interpreter"
+    DLE_ArrayZip _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr1, V_Array arr2) -> do
+          return $ V_Array $ map (\(l,r)-> V_Tuple $ [l,r]) $ zip arr1 arr2
+        _ -> impossible "expression interpreter"
+    DLE_TupleRef _at dlarg n -> do
+      ev <- interp dlarg
+      case ev of
+        V_Tuple arr -> return $ arr !! (fromIntegral n)
+        _ -> impossible "expression interpreter"
+    DLE_ObjectRef _at dlarg str -> do
+      ev <- interp dlarg
+      case ev of
+        V_Object obj -> return $ obj M.! str
+        _ -> impossible "expression interpreter"
+    DLE_Interact _at _slcxtframes _slpart _str _dltype dlargs -> do
+      args <- mapM interp dlargs
+      suspend $ PS_Suspend (Action_Interact args)
```

Comment:
You need to know the function name and the type. It might be helpful to show `at` and the frames too for a beautiful debugging experience